### ISSUE TITLE
[BUG] Build of .NET Core 3.1 project fails with: IAsyncEnumerable exists in both Microsoft.Bcl.AsyncInterfaces and System.Runtime

### ISSUE THREAD

**Issue by zmarty** [Azure/azure-sdk-for-net#9632]
**Describe the bug**
We are internal Microsoft team. We have .NET Core 3.1 projects which depends on Azure libraries that in turn depend on Azure.Core, which in turn references Microsoft.Bcl.AsyncInterfaces.

Microsoft.Bcl.AsyncInterfaces implements IAsyncEnumerable, which is now included natively in .Net Core 3.1.

The build fails since IAsyncEnumerable is present both in Microsoft.Bcl.AsyncInterfaces and in System.Runtime.

error CS0433: The type 'IAsyncEnumerable' exists in both 'Microsoft.Bcl.AsyncInterfaces, Version=1.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' and 'System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'

**Expected behavior**
What is the expected behavior?

Azure Pipelines build works

**Actual behavior (include Exception or Stack Trace)**
What is the actual behavior?

Azure Pipelines build does not work - see internal log https://msasg.visualstudio.com/Bing_UX/_build/results?buildId=9345941&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=7510ef29-db76-5a4c-4d87-8cd81d6b33b8&l=135

**To Reproduce**
Steps to reproduce the behavior (include a code snippet, screenshot, or any additional information that might help us reproduce the issue)

1. Create a project with .NET Core 3.1
2. Add references to Azure libraries, which themselves reference Azure.Core 1.0.1
3. Write a class in the new project that uses IAsyncEnumerable<T>
4. Have Azure Pipelines VSTS build the project (project builds and runs fine in Visual Studio 2019 16.4.1)

**Environment:**
 - Name and version of the Library package used: [e.g. Azure.Storage.Blobs 12.2.0] 
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): [e.g. Azure AppService or Windows 10 .NET Framework 4.8]
 - IDE and version : [e.g. Visual Studio 16.3]

Azure Pipelines .NET Core 3.1

Related bug: https://github.com/microsoft/azure-pipelines-tasks/issues/12209

**Comment by jsquire** [comment]
//cc: @pakrym, @AlexGhiondea

**Comment by zmarty** [comment]
Before you ask, I did try working around this with a using alias, but that does not work.

**Comment by zmarty** [comment]
Worked with @pakrym to fix this for us. Here is some information that may help others that get into the same situation:
- The reason this failure only happened in the Azure Pipelines build is that our pipeline used to restore nuget packages using an older version of nuget (4.7.1).
- It seems that never versions of nuget, as well as the build-in dotnet restore command, contain some magic code which omits the Microsoft.Bcl.AsyncInterfaces package from the restore tree, since it detects that we are using a ,NET Core 3.1

So in conclusion, either use a newer version of nuget in the pipeline, or better yet, use the dotnet restore command.
